### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by YauheniAntsipenka**
image
image

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,46 +1,54 @@
...
+import React, { useEffect } from 'react';
 import { Navigate, Route, Routes, useNavigate } from 'react-router-dom';
+import { useDispatch, useSelector } from 'react-redux';
 
 import { Courses } from './components/Courses/Courses';
 import { Header } from './components/Header/Header';
-import { CourseCardProps } from './components/Courses/components/CourseCard/CourseCard.types';
 
-import './App.scss';
+import './app.scss';
```
Comment (Margaryta-Maletz) [inline src/App.tsx]:
```suggestion
import './App.scss';
```

File: src/components/CreateCourse/CreateCourse.tsx [hunk 2]
```diff
@@ -1,14 +1,23 @@
...
+import React, { useEffect, useState } from 'react';
 import { useNavigate } from 'react-router-dom';
 
 import { Button } from '../../common/Button/Button';
 import { Input } from '../../common/Input/Input';
 
-import './CreateCourse.scss';
-import { AuthorItemProps } from './components/AuthorItem/AuthorItem.types';
-import { CourseProps } from './CreateCourse.types';
+import './createcourse.scss';
```
Comment (Margaryta-Maletz) [inline src/components/CreateCourse/CreateCourse.tsx]:
```suggestion
import './CreateCourse.scss';
```

File: src/components/Header/Header.tsx [hunk 3]
```diff
@@ -5,7 +5,7 @@ import { Button } from '../../common/Button/Button';
 
 import { HeaderProps } from './Header.types';
 
-import './Header.scss';
+import './header.scss';
```
Comment (Margaryta-Maletz) [inline src/components/Header/Header.tsx]:
```suggestion
import './Header.scss';
```

File: src/components/Header/components/Logo/Logo.tsx [hunk 4]
```diff
@@ -2,6 +2,6 @@ import React from 'react';
 
 import mainLogo from '../../../../resources/logo.png';
 
-import './Logo.scss';
+import './logo.scss';
```
Comment (Margaryta-Maletz) [inline src/components/Header/components/Logo/Logo.tsx]:
```suggestion
import './Logo.scss';
```

File: src/components/Login/Login.tsx [hunk 5]
```diff
@@ -1,22 +1,31 @@
...
 import { Header } from '../Header/Header';
 import { Button } from '../../common/Button/Button';
 import { Input } from '../../common/Input/Input';
 
-import { LoginProps } from './Login.types';
-
-import './Login.scss';
-import '../../App.scss';
+import './login.scss';
+import '../../app.scss';
```
Comment (Margaryta-Maletz) [inline src/components/Login/Login.tsx]:
```suggestion
import '../../App.scss';
```

File: src/components/Registration/Registration.tsx [hunk 6]
```diff
@@ -1,22 +1,31 @@
...
 import { Header } from '../Header/Header';
 import { Button } from '../../common/Button/Button';
 import { Input } from '../../common/Input/Input';
 
-import { RegistrationProps } from './Registration.types';
-
-import './Registration.scss';
-import '../../App.scss';
+import './registration.scss';
+import '../../app.scss';
```
Comment (Margaryta-Maletz) [inline src/components/Registration/Registration.tsx]:
```suggestion
import './Registration.scss';
import '../../App.scss';
```

File: src/services.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+		headers: new Headers({ 'content-type': 'application/json' }),
+		body: JSON.stringify(data),
+	})
+		.then((response) => response.json())
+		.then((responseData) => {
+			localStorage.setItem('token', responseData.result);
+			localStorage.setItem('username', responseData.user.name);
+		})
+		.catch((error) => {
+			console.log(error);
```
Comment (Margaryta-Maletz) [inline src/services.ts:28]: When you catch error it is better to use console.error to show error
```suggestion
			console.error(error);
```

File: src/services.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+		.catch((error) => console.log(error));
+}
+
+export function getAllCourses() {
+	let courses: CourseType[] = [];
+	fetchCourses().then((coursesLocal) => {
+		console.log('coursesLocal: ', coursesLocal);
+		courses.concat(coursesLocal);
+	});
+	console.log('courses:', courses);
```
Comment (Margaryta-Maletz) [inline src/services.ts]: Do you need this console? Do not forget to delete redundant logs here and below in this file.

File: src/services.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+export async function fetchCourses(): Promise<CourseType[]> {
+	const response = await fetch('http://localhost:4000/courses/all', {
+		method: 'GET',
+		headers: new Headers({ 'content-type': 'application/json' }),
+	});
+
+	type JSONResponse = {
+		successful: boolean;
+		result?: CourseType[];
+	};
```
Comment (Margaryta-Maletz) [inline src/services.ts]: Oh, it's better to use one type for all responses. And here you can see a practical example of using generic.
1. Add a new type above with generic.
2. Delete the type here.
3. Specify a type variable when use this type.
```suggestion
```

File: src/services.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+	const { successful, result }: JSONResponse = await response.json();
```
Comment (Margaryta-Maletz) [inline src/services.ts]:
```suggestion
	const { successful, result }: JSONResponse<CourseType[]> =
		await response.json();
```
Make changes here and below

File: src/services.ts [hunk 11]
```diff
@@ -0,0 +1,185 @@
+import { LoginProps } from './components/Login/Login.types';
+import { RegistrationProps } from './components/Registration/Registration.types';
+import { CourseType } from './store/courses/types';
+import { AuthorType } from './store/authors/types';
+
```
Comment (Margaryta-Maletz) [inline src/services.ts:10]:
```suggestion

type JSONResponse<T> = {
	successful: boolean;
	result?: T;
};

```

File: src/services.ts [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+		successful: boolean;
+		result?: CourseType[];
+	};
+
+	const { successful, result }: JSONResponse = await response.json();
+
+	if (successful === true && result?.length) {
+		return result;
+	}
+	return [] as CourseType[];
```
Comment (Margaryta-Maletz) [inline src/services.ts]: 'if' statement can be simplified. For example
```suggestion
	return successful && result?.length ? result : [] as CourseType[];
```
Try to make the same changes below

File: src/services.ts [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+	});
+
+	type JSONResponse = {
+		successful: boolean;
+		result?: CourseType;
+	};
+
+	const { successful, result }: JSONResponse = await response.json();
+	console.log(data);
+	console.log(result);
```
Comment (Margaryta-Maletz) [inline src/services.ts]: Now it is not working for me :(
image

image

File: src/services.ts [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+	if (successful === true) {
+		return true;
+	}
+	return false;
```
Comment (Margaryta-Maletz) [inline src/services.ts]: 'if' statement can be simplified
```suggestion
	return successful;
```

**Review by Margaryta-Maletz — CHANGES_REQUESTED** [review]
Generally, good job!
But I'm having some problems with the functionality of it
image
Could you please clarify these points without tick?

What's wrong? If token in the localStorage the user should be immediately redirected to the `Courses` page when opening the application.
image

Oh, I'm still seeing these errors. In the current branch, you changed file and link names at the same time. It is mistake.
image

You can check the file names on each branch.
image

File: src/App.tsx (hunk 1 above)
Comment (YauheniAntsipenka) [reply src/App.tsx]: all scss files were deleted and recreated in separate commits

File: src/components/CreateCourse/CreateCourse.tsx (hunk 2 above)
Comment (YauheniAntsipenka) [reply src/components/CreateCourse/CreateCourse.tsx]: all scss files were deleted and recreated in separate commits

File: src/components/Header/Header.tsx (hunk 3 above)
Comment (YauheniAntsipenka) [reply src/components/Header/Header.tsx]: all scss files were deleted and recreated in separate commits

File: src/components/Header/components/Logo/Logo.tsx (hunk 4 above)
Comment (YauheniAntsipenka) [reply src/components/Header/components/Logo/Logo.tsx]: all scss files were deleted and recreated in separate commits

File: src/components/Login/Login.tsx (hunk 5 above)
Comment (YauheniAntsipenka) [reply src/components/Login/Login.tsx]: all scss files were deleted and recreated in separate commits

File: src/components/Registration/Registration.tsx (hunk 6 above)
Comment (YauheniAntsipenka) [reply src/components/Registration/Registration.tsx]: all scss files were deleted and recreated in separate commits

File: src/services.ts (hunk 7 above)
Comment (YauheniAntsipenka) [reply src/services.ts:28]: but there are no logs here in pr 
image

File: src/services.ts (hunk 8 above)
Comment (YauheniAntsipenka) [reply src/services.ts]: thanks, removed

File: src/services.ts (hunk 9 above)
Comment (YauheniAntsipenka) [reply src/services.ts]: thanks, refactored

File: src/services.ts (hunk 11 above)
Comment (YauheniAntsipenka) [reply src/services.ts:10]: thanks, refactored

File: src/services.ts (hunk 12 above)
Comment (YauheniAntsipenka) [reply src/services.ts]: thanks, refactored

File: src/services.ts (hunk 14 above)
Comment (YauheniAntsipenka) [reply src/services.ts]: thanks, refactored

**Pull request merged**